File: c++/linked_list.cpp
#include<iostream>
using namespace std;

struct node {
    int data;
    struct node *next;
};
typedef struct node node;

node* insert_begin(node *n, int num) {
    node *tmp = new node();
    tmp->data = num;
    tmp->next = n;
    cout<<"Inserted "<<num<<" at the begining\n";
    return tmp;
}

void insert_end(node *n, int num) {
    node *tmp = new node();
    tmp->data = num;
    tmp->next = nullptr;
    for (node *curr = n; curr != nullptr; curr = curr->next) {
        if (curr->next == nullptr) {
            curr->next = tmp;
            cout<<"Inserted "<<num<<" at the end\n";
            break;
        }
    }
}

void insert_middle(node *n, int num, int pos) {
    node *tmp = new node();
    tmp->data = num;
    int i = 1;
    for (node *curr = n; curr != nullptr; curr = curr->next) {
        if (i == pos) {
            tmp->next = curr->next;
            curr->next = tmp;
            cout<<"Inserted "<<num<<" after "<<pos<<" node\n";
            break;
        }
        i++;
    }
}

node* delete_begin(node *n) {
    cout<<"Deleted "<<n->data<<endl;
    node *tmp = n->next;
    return tmp;
}

void delete_end(node *n) {
    for (node *curr = n; curr != nullptr; curr = curr->next) {
        if(curr->next->next == nullptr) {
            cout<<"Deleted "<<curr->next->data<<endl;
            curr->next = nullptr;
            break;
        }
    }
}

void delete_middle(node *n, int pos) {
    int i = 2;
    for (node *curr = n; curr != nullptr; curr = curr->next) {
        if (i == pos) {
            cout<<"Deleted "<<curr->next->data<<endl;
            curr->next = curr->next->next;
            break;
        }
        i++;
    }
}

void traverse(node *n) {
    cout<<"Elements of Linked List"<<endl;
    for (node *curr = n; curr != nullptr; curr = curr->next) {
        cout<<curr->data<<endl;
    }
}

int main() {
    node *head = nullptr;
    head = insert_begin(head, 1);
    head = insert_begin(head, 7);
    insert_end(head, 3);
    insert_middle(head, 4, 2);
    traverse(head);
    head = delete_begin(head);
    delete_middle(head, 2);
    delete_end(head);
    traverse(head);
}

File: java/LinkedList.java
class LinkedList {
    static class Node {
        int value;
        Node next;

        Node(int d) {
            value = d;
            next = null;
        }
    }
    Node head;

    public void insert_begin(int num) {
        if (head == null) {
            head = new Node(num);
            return;
        }
        Node tmp = new Node(num);
        tmp.next = head;
        head = tmp;
    }

    public void insert_end(int num) {
        if (head == null) {
            head = new Node(num);
            return;
        }
        for(Node tmp = head; tmp != null; tmp = tmp.next) {
            if (tmp.next == null) {
                tmp.next = new Node(num);
                return;
            }
        }
    }

    public void insert_middle(int num, int pos) {
        int i = 1;
        if (pos == 0) {
            insert_begin(num);
            return;
        }
        for(Node curr = head; curr != null; curr = curr.next) {
            if (i == pos) {
                Node tmp = new Node(num);
                tmp.next = curr.next;
                curr.next = tmp;
                return;
            }
            i++;
        }
    }

    public void delete_begin() {
        head = head.next;
    }

    public void delete_end() {
        if (head.next == null) {
            head = null;
            return;
        }
        for (Node curr = head; curr != null; curr = curr.next) {
            if (curr.next.next == null) {
                curr.next = null;
            }
        }
    }

    public void delete_middle(int pos) {
        int i = 1;
        for (Node curr = head; curr != null; curr = curr.next) {
            if (i == pos - 1) {
                curr.next = curr.next.next;
                break;
            }
            i++;
        }
    }

    public void traverse() {
        for (Node curr = head; curr != null; curr = curr.next) {
            System.out.print(curr.value+"->");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.insert_begin(7);
        ll.insert_begin(1);
        ll.insert_begin(2);
        ll.traverse();
        ll.insert_end(3);
        ll.insert_end(4);
        ll.traverse();
        ll.insert_middle(8, 0);
        ll.insert_middle(8, 4);
        ll.traverse();
        ll.delete_begin();
        ll.traverse();
        ll.delete_end();
        ll.traverse();
        ll.delete_middle(3);
        ll.traverse();
    }
}
